aws/resource_aws_glue_dev_endpoint.go

```diff
@@ -0,0 +1,580 @@
+	}
+
+	return nil
+}
+
+func resourceAwsDevEndpointUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).glueconn
+
+	input := &glue.UpdateDevEndpointInput{
+		EndpointName: aws.String(d.Get("name").(string)),
+	}
+
+	hasChanged := false
+
+	customLibs := &glue.DevEndpointCustomLibraries{}
+
+	if d.HasChange("arguments") {
+		oldRaw, newRaw := d.GetChange("arguments")
+		old := oldRaw.(map[string]interface{})
+		new := newRaw.(map[string]interface{})
+		create, remove := diffArguments(old, new)
+		input.AddArguments = create
+		input.DeleteArguments = remove
+
+		hasChanged = true
+	}
+
+	if d.HasChange("extra_jars_s3_path") {
+		customLibs.ExtraJarsS3Path = aws.String(d.Get("extra_jars_s3_path").(string))
+		input.CustomLibraries = customLibs
+		input.UpdateEtlLibraries = aws.Bool(true)
+
+		hasChanged = true
+	}
+
+	if d.HasChange("extra_python_libs_s3_path") {
+		customLibs.ExtraPythonLibsS3Path = aws.String(d.Get("extra_python_libs_s3_path").(string))
+		input.CustomLibraries = customLibs
+		input.UpdateEtlLibraries = aws.Bool(true)
+
+		hasChanged = true
+	}
+
+	if d.HasChange("public_key") {
+		input.PublicKey = aws.String(d.Get("public_key").(string))
+
+		hasChanged = true
+	}
+
+	if d.HasChange("public_keys") {
+		o, n := d.GetChange("public_keys")
+		if o == nil {
+			o = new(schema.Set)
+		}
+		if n == nil {
+			n = new(schema.Set)
+		}
+		os := o.(*schema.Set)
+		ns := n.(*schema.Set)
+		remove := os.Difference(ns).List()
+		create := ns.Difference(os).List()
+
+		input.AddPublicKeys = expandStringList(create)
+		log.Printf("[DEBUG] expectedCreate public keys: %v", create)
+
+		input.DeletePublicKeys = expandStringList(remove)
+		log.Printf("[DEBUG] remove public keys: %v", remove)
+
+		hasChanged = true
+	}
+
+	if hasChanged {
+		log.Printf("[DEBUG] Updating Glue Dev Endpoint: %s", input)
+
+		_, err := conn.UpdateDevEndpoint(input)
+		if err != nil {
+			return fmt.Errorf("error updating Glue Dev Endpoint: %s", err)
+		}
+	}
+
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
+		if err := keyvaluetags.GlueUpdateTags(conn, d.Get("arn").(string), o, n); err != nil {
+			return fmt.Errorf("error updating tags: %s", err)
+		}
+	}
+
+	return resourceAwsGlueDevEndpointRead(d, meta)
+}
+
+func resourceAwsDevEndpointDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).glueconn
+
+	deleteOpts := &glue.DeleteDevEndpointInput{
+		EndpointName: aws.String(d.Id()),
+	}
+
+	log.Printf("[INFO] Deleting Glue Dev Endpoint: %s", d.Id())
+
+	_, err := conn.DeleteDevEndpoint(deleteOpts)
+	if err != nil {
+		if isAWSErr(err, glue.ErrCodeEntityNotFoundException, "") {
+			return nil
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+func diffArguments(oldArgs, newArgs map[string]interface{}) (map[string]*string, []*string) {
+	var create = make(map[string]*string)
+	var remove []*string
+
+	for oldArgKey, oldArgVal := range oldArgs {
```

Comment:
instead of for in for:

pseudocode:

```
for oldK, oldV in old_map {
  if val, ok := new_map[oldK]; ok {
    if val != oldV {
      removed = append(removed, old)
    }
  } else {
   removed = append(removed, old)
}

for newK, newV in new_map {
  if _, ok := dict[newK]; !ok {
    create[newK] = newV
  }
}
```